# Matrix multiplication.
# Understanding Usage of multiplication of matrix and it's application.
# Checking for scores of movies

# We take two matrix one for users/viewers WITH 3 genres with user like scores,

# and one for movies with 3 different genres.



In [24]:
import numpy as np

In [25]:
#Cols represent preferences for: `[Action, Comedy, Sci-Fi]`.
#rows represent viewers
'''users=np.array([
    [1,0,5],
    [0,0,5],
    [2,3,3]]
          )
#Cols represent specific movies: `[GGVV, UK, Kantara]`.
#rows represent generes
movies=np.array([
    [1,1,1],
    [0,0,1],
    [1,0,1]]
           )'''
import numpy as np

# FIX: Add a dot '.' to make them floats (1. instead of 1)
# Or use dtype=float
users = np.array([
    [1., 0., 5.],
    [0., 0., 5.],
    [2., 3., 3.]
])

movies = np.array([
    [1., 1., 1.],
    [0., 0., 1.],
    [1., 0., 0.]
])

res=np.dot(users,movies)
movie_names = ["GGVV", "UK", "Kantara"]

print('GGVV | UK | Kantara')
print(f"{res[0]}")
print(f"{res[1]}")
print(f"{res[2]}")

#3. THE LOGIC: Identify which movies are liked
# We loop through each user to interpret their scores
for i in range(len(res)):
    user_id = i + 1
    user_scores = res[i]
    
    print(f"\n--- Analysis for User {user_id} ---")
    
    # Strategy: Find the highest score (Favorite Movie)
    best_score_index = np.argmax(user_scores) # Returns index of highest number
    favorite_movie = movie_names[best_score_index]
    
    print(f"Favorite Movie: {favorite_movie}")
    
    # Strategy: List all 'Liked' movies (Threshold > 5)
    print("All Liked Movies (Score > 5):")
    for j in range(len(user_scores)):
        if user_scores[j] > 5:
            print(f" - {movie_names[j]} (Score: {user_scores[j]})")

GGVV | UK | Kantara
[6. 1. 1.]
[5. 0. 0.]
[5. 2. 5.]

--- Analysis for User 1 ---
Favorite Movie: GGVV
All Liked Movies (Score > 5):
 - GGVV (Score: 6.0)

--- Analysis for User 2 ---
Favorite Movie: GGVV
All Liked Movies (Score > 5):

--- Analysis for User 3 ---
Favorite Movie: GGVV
All Liked Movies (Score > 5):


# Adding "Bias"

## 1. The Concept: Why "+ b"?
In the real world, "Similarity" isn't the only factor.

User Bias: Some users are "Critics" (rate everything lower).
Some are "Optimists" (rate everything higher).

Item Bias: Some movies are just universally popular, regardless of genre.

To model this, we change our equation from linear (y = ax) to affine (y = ax + b).
Final Score = User (dot) Movie + Bias

## 2. The Implementation: Broadcasting
We want to add a specific "Mood Score" to each user.
* **User 0 (Critic):** Subtracts 2 points from *every* movie.
* **User 1 (Optimist):** Adds 2 points to *every* movie.
* **User 2 (Neutral):** Adds 0 points.

**The Broadcasting Challenge:**
* `Scores` Matrix shape: `(3 Users, 3 Movies)`
* `Bias` Vector shape: `(3 Users)`
To add them, we must reshape the Bias vector to `(3, 1)` (Vertical). This forces NumPy to apply the bias **Row-by-Row** (per user) instead of Column-by-Column.

In [26]:
#mood of users
#user_bias=np.array([-2,0,1])
# Re-define bias as floats too
user_bias = np.array([-2., 2., 0.])
user_biass=user_bias.reshape(-1,1)

Res=res+user_biass
print(Res)

# I can use the same logic again to find user and movie analysis

[[ 4. -1. -1.]
 [ 7.  2.  2.]
 [ 5.  2.  5.]]


When we eventually use PyTorch or TensorFlow, you will see layers defined like Linear(in_features=3, out_features=1, bias=True)

this is the same logic we built from scratch.


# Activation Function (ReLU)

## 1. The Concept
Our linear calculation ($Y = WX + b$) is just math. It can produce values from $-\infty$ to $+\infty$.
But biological neurons don't send "negative signals." They either **Fire** (send a signal) or **Stay Silent** (0).

To mimic this, and to introduce **Non-Linearity** (which allows AI to learn complex patterns), we use an Activation Function.

## 2. The ReLU (Rectified Linear Unit)
It is the most common activation in modern AI (ChatGPT, Computer Vision).
$$f(x) = \max(0, x)$$

* **Input:** `[-2, 5, -8, 10]`
* **Output:** `[ 0, 5,  0, 10]`

In our Recommender:
* Negative scores (Dislikes) become **0** (Ignored).
* Positive scores (Likes) remain **Linear** (Stronger likes = Higher numbers).

In [27]:
# np.maximum compares 0 and the score, picking the larger one.
activated_scores = np.maximum(0, Res)

print("\n--- Final Output after ReLU (Activation) ---")
print(activated_scores)


--- Final Output after ReLU (Activation) ---
[[4. 0. 0.]
 [7. 2. 2.]
 [5. 2. 5.]]


I will put the analysis code again on final output.

In [28]:
# 3. THE LOGIC: Identify which movies are liked
# We loop through each user to interpret their scores
for i in range(len(activated_scores)):
    user_id = i + 1
    user_scores = activated_scores[i]
    
    print(f"\n--- Analysis for User {user_id} ---")
    
    # Strategy: Find the highest score (Favorite Movie)
    best_score_index = np.argmax(user_scores) # Returns index of highest number
    favorite_movie = movie_names[best_score_index]
    
    print(f"Favorite Movie: {favorite_movie}")
    
    # Strategy: List all 'Liked' movies (Threshold > 5)
    print("All Liked Movies (Score > 5):")
    for j in range(len(user_scores)):
        if user_scores[j] > 5:
            print(f" - {movie_names[j]} (Score: {user_scores[j]})")


--- Analysis for User 1 ---
Favorite Movie: GGVV
All Liked Movies (Score > 5):

--- Analysis for User 2 ---
Favorite Movie: GGVV
All Liked Movies (Score > 5):
 - GGVV (Score: 7.0)

--- Analysis for User 3 ---
Favorite Movie: GGVV
All Liked Movies (Score > 5):


Summary 


##Forward Pass:
Definition: Calculating the prediction ($Y$) from the input ($X$).
Math: $Y = \text{Activation}(X \cdot W + b)$Your Code: activated_scores = np.maximum(0, (np.dot(users, movies) + bias))


##Dense Layer:
Definition: A layer of neurons where each neuron receives input from all neurons in the previous layer.Implementation: It is just Matrix Multiplication.Why "Dense"? Because the web of connections is "dense" (full), not sparse (empty)

Gemini Guide.

# 📉 The Loss Function (MSE)

## 1. The Goal
We built a model that *guesses* scores. Now we need to measure accuracy.
We cannot "learn" if we don't know how big our mistake is.

## 2. The Math: Mean Squared Error (MSE)
$$L = \frac{1}{n} \sum (y_{true} - y_{pred})^2$$

* **Why Square it?**
    * If we predict 8 but truth is 10, error is **-2**.
    * If we predict 12 but truth is 10, error is **+2**.
    * If we just add them, $-2 + 2 = 0$. The model thinks it's perfect!
    * Squaring makes both errors **4**, so the total error is **8**. It forces the model to treat *all* mistakes as bad.

## 3. The Result
* **Loss:** A single number that summarizes the "Total Stupidity" of the model.
* **Goal of AI:** Change the Weights (Movies Matrix) until this number becomes **0**.

In [29]:
# 1. THE REALITY (Ground Truth)
# Let's say these are the ACTUAL ratings the users gave
# User 0 rated Movie 0 as a 10. User 1 rated Movie 1 as a 5.
y_true = np.array([
    [10, 0, 10], # User 0 Loved GGVV and Kantara
    [0, 5, 0],   # User 1 Loved UK
    [2, 2, 2]    # User 2 was just "meh" on everything
])

# 2. OUR PREDICTION (From your previous code)
# Let's pretend 'activated_scores' is what our model just guessed
# (In a real script, this comes from the Forward Pass)
y_pred = activated_scores 

# 3. CALCULATE THE LOSS (Mean Squared Error)
# Math: Average of (True - Predicted)^2
error = y_true - y_pred
squared_error = error ** 2
mse = np.mean(squared_error)

print("--- The Scorecard ---")
print("True Ratings:\n", y_true)
print("\nPredicted Ratings:\n", y_pred)
print(f"\nMean Squared Error (Loss): {mse:.4f}")

--- The Scorecard ---
True Ratings:
 [[10  0 10]
 [ 0  5  0]
 [ 2  2  2]]

Predicted Ratings:
 [[4. 0. 0.]
 [7. 2. 2.]
 [5. 2. 5.]]

Mean Squared Error (Loss): 24.0000


the mse is high so our predictions are worse.


now to lessen the error or loss we have two options
1. Manually changing the values of movie matrix and check for decrese in loss(this takes undefineded steps).
2. Calculus[Gradient]:Using Gradient tells us  which direction to nudge to lower the loss.
 Visual Map:

>Imagine you are on top of a mountain (High Error).

>You want to get to the bottom of the valley (Zero Error).

>The Loss Function tells you your current height.

>The Gradient tells you which way is "downhill".

>The Learning Rate is how big of a step you take.

OKAY BEFORE GOING FOR GRADIENT I WILL TRY CHNGING MANUALLY AND CHECK FOR LOSS.

In [30]:
# 1. CHECK BEFORE
# Recalculate current loss
current_pred = np.maximum(0, np.dot(users, movies) + user_bias.reshape(-1,1))
loss_before = np.mean((y_true - current_pred) ** 2)

print(f"Loss BEFORE update: {loss_before:.4f}")
print(f"User 0 Rating for GGVV (Before): {current_pred[0][0]}")

# 2. THE MANUAL "LEARNING" STEP
# We know User 0's rating (6) is too low (Target is 10).
# User 0 has a '1' in Action (index 0). 
# Let's INCREASE the 'Action' value for Movie 0 (GGVV).

# movies[0, 0] represents: Row 0 (Action), Col 0 (GGVV)
print("\n--- Updating Weight: Increasing Action score for GGVV ---")
movies[0, 0] += 0.5  # Nudging it up from 1.0 to 1.5

# 3. CHECK AFTER
new_pred = np.maximum(0, np.dot(users, movies) + user_bias.reshape(-1,1))
loss_after = np.mean((y_true - new_pred) ** 2)

print(f"User 0 Rating for GGVV (After): {new_pred[0][0]}")
print(f"Loss AFTER update: {loss_after:.4f}")

if loss_after < loss_before:
    print("\n✅ SUCCESS: The model learned! Loss went down.")
else:
    print("\n❌ FAIL: We moved the wrong way.")

Loss BEFORE update: 24.0000
User 0 Rating for GGVV (Before): 4.0

--- Updating Weight: Increasing Action score for GGVV ---
User 0 Rating for GGVV (After): 4.5
Loss AFTER update: 24.1389

❌ FAIL: We moved the wrong way.


Why did it fail?You fixed the problem for User 0, but you accidentally broke it for User 2.
Let's look at the math trace:

User 0 (Action Lover):
Target: 10Old 
Prediction: 4.0 (Error: $36$)
New Prediction: 4.5 (Error: $30.25$)
Result: Improved (Error went down by 5.75).

User 2 (The "Meh" Guy):His input is [2, 3, 3] (He likes Action a little bit).Target: 2 (He hated GGVV).
Old Prediction: 5.0. (Already too high!)
New Prediction: 6.0. (Because you increased the Action score, his prediction went even higher).
Result:  Worsened (Error went from 9 up to 16).
The Net Result:The "damage" to User 2 (Error +7) was bigger than the "help" to User 0 (Error -5.75). So the Total Loss went up.

The Lesson: This isn't about making one person happy. It's about finding the perfect compromise that makes the least amount of people unhappy.User 0 wants the weight to go UP.User 2 wants the weight to go DOWN.Manual tuning is impossible because you have thousands of these conflicts. You can't calculate them all in your head.🚀 

The Solution: Automation (Calculus)This is exactly why we need Calculus.Instead of us guessing "+0.5", we calculate the Derivative. The Derivative looks at User 0, User 1, and User 2 simultaneously and tells us:"If you move up, User 0 gains 5 points but User 2 loses 7 points. Therefore, the smart move is to go DOWN."

Phase 1 Complete.
You have now successfully:

Built the Data Structure (Matrices).

Built the Forward Pass (Dot Product + Bias + ReLU).

Calculated the Error (MSE).

Attempted Backpropagation (and learned why it's hard).